### DISCUSSION TITLE
Dynamic OpenAPI (Swagger/ReDoc) documents based on route prefix

### DISCUSSION THREAD

**Discussion by lccmrx** [fastapi/fastapi#6529] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI

prefix = "v1" # But it could be v2, v3, and so on.

app = FastAPI(
        title="Content Manager",
        openapi_url=f"{prefix}/docs/openapi.json",
        docs_url=f"/{prefix}/docs/",
        redoc_url=f"/{prefix}/docs/redoc",
    )
```

### Description

So I'm writing a API that's surely going to scale through versioning of my routes.
Since FastAPI offers the OpenAPI integration so beautifuly, I wondered if my prefix versions could show only the document of the "route prefix" version.

I've tried to instantiate dynamically as inputted, but as the prefix is a fixed variable, it could not get it.

Has anybody achieved some thing sort of equal to what I'm wanting?

### Operating System

macOS

### FastAPI Version

0.68.0

### Python Version

3.9.5

**Comment by Dustyposa** [comment]
Did you want to change `docs prefix` in running server?
If not, you can do :
```python
from fastapi import FastAPI

prefix = get_pre_fix() # But it could be v2, v3, and so on.

app = FastAPI(
        title="Content Manager",
        openapi_url=f"/{prefix}/docs/openapi.json",
        docs_url=f"/{prefix}/docs/",
        redoc_url=f"/{prefix}/docs/redoc",
    )
```

**Comment by lccmrx** [comment]
Hi @Dustyposa !

Yes, I would like to do that. But using the custom_open_api like so:
```python
def custom_openapi():
     openapi_schema = get_openapi(
         description="This is a very custom OpenAPI schema",
         routes= # Prefixed routes
         # If v1 prefix is selected, it would only show the v1 routes.
         # If v2 prefix is selected, it would only show the v2 routes.
     )
     api.openapi_schema = openapi_schema
     return api.openapi_schema

api.openapi = custom_openapi
```

**Comment by Kludex** [comment, marked as answer]
You can follow this: https://github.com/tiangolo/fastapi/issues/381#issuecomment-893965341 . Meaning that you'd have to create an app for each version and mount them.

Or you can do this: https://gist.github.com/Kludex/1c515aa38d22ec28d514da6b6f36da9f

**Comment by lccmrx** [comment]
Oh! That's nice @Kludex! That's perfect btw!

Thanks for the help guys!

**Comment by lccmrx** [comment]
@Kludex, I've run into an issue with this method.

On my main app (before applying your method) I had a startup function that created some resources in app.state on FastAPI.

After trying your method out, it seems like my startup method only works based on the main app, so I tried to add the same startup method on the versioned apps. But somehow it still fails, saying my object inside state weren't defined.

Feel free to give it a look on what it looks like.
```python
def create_startup_handler(app: FastAPI):
    def startup():
        app.state.database_provider = DatabaseProvider()
        app.state.http_provider = HttpProvider()
        app.state.auth_provider = AuthProvider()
        app.state.cache_provider = CacheProvider()

    return startup

def create_shutdown_handler(app: FastAPI):
    async def shutdown():
        await app.state.http_provider.finish()

    return shutdown

def get_application():
    app = FastAPI(
        title="Content Manager",
        openapi_url=f"/docs/api-description.json" if DOCS_ENABLED else "",
        docs_url=f"/docs/",
        redoc_url=f"/docs/redoc",
        openapi_tags=[
            {
                'name': 'v1',
                "description": "API version 1, check link on the right",
                'externalDocs': {
                    'description': 'sub-docs',
                    'url': 'http://127.0.0.1:5000/api/v1/docs'
                }
            },
        ]
    )
    include_global_middlewares(app)

    app.add_event_handler("startup", create_startup_handler(app))
    app.add_event_handler("shutdown", create_shutdown_handler(app))
    set_api_error_handler(app)
    
    create_app_versions(app, "v1", v1_router)
    
    return app

def include_global_middlewares(app: FastAPI):
    app.add_middleware(
        cors.CORSMiddleware,
        allow_origins=["*"],
        allow_credentials=False,
        allow_methods=["*"],
        allow_headers=["*"],
    )
    (lambda x: app.add_middleware(httpsredirect.HTTPSRedirectMiddleware) if x else None) (HTTPS_REDIRECT)
    app.add_middleware(
        trustedhost.TrustedHostMiddleware, allowed_hosts=ALLOWED_HOSTS
    )

def create_app_versions(main_app: FastAPI, version: str, router):
    app_version = FastAPI(
        title=f"Content Manager - {version}",
        openapi_url=f"/docs/api-description.json" if DOCS_ENABLED else "",
        docs_url=f"/docs/",
        redoc_url=f"/docs/redoc"
    )
    include_global_middlewares(app_version)

    app_version.add_event_handler("startup", create_startup_handler(app_version))
    app_version.add_event_handler("shutdown", create_shutdown_handler(app_version))
    set_api_error_handler(app_version)
    
    app_version.include_router(router)
    main_app.mount(f"/api/{version}", app_version)

app = get_application()

def main():
    import uvicorn
    
    uvicorn.run("content_manager.main:app", port=PORT, reload=True)
```

**Comment by dust63** [comment]
```

 'externalDocs': {
                    'description': 'sub-docs',
                    'url': 'http://127.0.0.1:5000/api/v1/docs'
                }
```

Do you try your implementation outside your localhost. I think you will have some trouble ;)

**Comment by lccmrx** [comment]
Hi @dust63!
I've created a good way using @Kludex method.
Have a look [here](https://github.com/DevLucca/python-fastapi-template)
